### ISSUE TITLE
Calling `HTTPure.Body.toString` multiple times on the same Readable stream blocks forever

### ISSUE THREAD

**Issue by boygao1992** [cprussin/purescript-httpure#188]
`body` was `String` in `Request` before `0.13.0` which could be consumed multiple times in different middlewares. After `0.13.0`, `body` is now `Readable ()` stream which cannot be cloned easily like a `String`, and if a stream is drained by a `HTTPure.Body.toString` once, the second `HTTPure.Body.toString` on the same stream will block forever. This can be verified in a simple test like the following
```purescript
    it "call `toString` multiple times on the same stream" do
      let
        stream = stringToStream "foobar"
      string1 <- toString stream
      string2 <- toString stream
      string1 ?= "foobar"
      string2 ?= "foobar"
```

`HTTPure.Body.toString` probably has to be stateful which caches and returns the result from the first run.

**Comment by cprussin** [comment]
Hmm this is interesting -- could be solved by changing the monad in which the request runs from just an `Aff` to a stack which includes `StateT`.  Any other ideas?

**Comment by boygao1992** [comment]
`StateT` is not reliable against runtime exceptions so I prefer `Ref`.

**Comment by cprussin** [comment]
@boygao1992 published v0.14.0 containing this fix.  Thanks again for the contributions.

cc @sigma-andex

**Comment by sigma-andex** [comment]
very cool! I'm wondering though, with this implementation it still wouldn't be possible to use the stream in the middleware  without consuming it entirely. Say I want to post a 100mb file and want to do some filtering, threat detection or whatever generalized in the middleware.

**Comment by sigma-andex** [comment]
but yeah I guess in this case one could just pass the request with the filtered stream to the next  handler..

**Comment by boygao1992** [comment]
I know the PR I submitted only recovers the old behavior consuming body as a `String` which is probably 100% the case in my code base. It is still possible that one drains the stream with the `Readable` stream interface outsides `toBuffer`/`toString` which the `RequestBody` state doesn't track, thus future `toBuffer`/`toString` will block forever still. To properly handle the interactions around middleware and stream, I think the library shouldn't give user direct access to the native `Readable` stream. Instead provide a set of primitives like `toBuffer`/`toString` for stream so that the `RequestBody` state can keep track of things like subscriptions, transformations, etc. and then properly clone/broadcast the stream with library like [cloneable-readable](https://www.npmjs.com/package/cloneable-readable) and [pumpify](https://www.npmjs.com/package/pumpify). Or if we want something fancier, we could wrap the `Readable` stream with a stream combinator interface like `purescript-run-streaming`, `purescript-pipes` but I had little experience using those so not sure about whether they are stack-safe and in which case one could introduce memory leak.

A side note on the response side, I attempted to add a middleware for compressing the response body with gzip but it seems like currently once the response body is turned into a `Response` with `Response.ok` or friends, the response body is no longer accessible as it's partially applied to [Body.write](https://github.com/cprussin/purescript-httpure/blob/bc93c23f21bb1ad844d29060f27aacbd4636cf05/src/HTTPure/Response.purs#L249).
I'm curious if there's an easy way to keep response body around so that middlewares can transform it. The solution I could think of right now might have to deprecate `Body` type class and gather all body types the library supports into a closed sum type.
